tests/proptest.rs

```diff
@@ -0,0 +1,436 @@
+                    );
+                }
+
+                (solver, complicated)
+            },
+        )
+}
+
+/// This test is to test the generator to ensure
+/// that it makes registries with large dependency trees
+#[test]
+fn meta_test_deep_trees_from_strategy() {
+    use proptest::strategy::ValueTree;
+    use proptest::test_runner::TestRunner;
+
+    let mut dis = [0; 21];
+
+    let strategy = registry_strategy(0u16..665, 666);
+    let mut test_runner = TestRunner::deterministic();
+    for _ in 0..128 {
+        let (dependency_provider, cases) = strategy
+            .new_tree(&mut TestRunner::new_with_rng(
+                Default::default(),
+                test_runner.new_rng(),
+            ))
+            .unwrap()
+            .current();
+
+        for (name, ver) in cases {
+            let res = resolve(&dependency_provider, name, ver);
+            dis[res
+                .as_ref()
+                .map(|x| std::cmp::min(x.len(), dis.len()) - 1)
+                .unwrap_or(0)] += 1;
+            if dis.iter().all(|&x| x > 0) {
+                return;
+            }
+        }
+    }
+
+    panic!(
+        "In {} tries we did not see a wide enough distribution of dependency trees! dis: {:?}",
+        dis.iter().sum::<i32>(),
+        dis
+    );
+}
+
+proptest! {
+    #![proptest_config(ProptestConfig {
+    max_shrink_iters:
+        if std::env::var("CI").is_ok() {
+            // This attempts to make sure that CI will fail fast,
+            0
+        } else {
+            // but that local builds will give a small clear test case.
+            2048 // u32::MAX
+        },
+        result_cache: prop::test_runner::basic_result_cache,
+        .. ProptestConfig::default()
+    })]
+
+    #[test]
+    /// This test is mostly for profiling
+    fn prop_passes_string(
+        (dependency_provider, cases) in registry_strategy(string_names(), "bad".to_owned())
+    )  {
+        for (name, ver) in cases {
+            let _ = resolve(&TimeoutDependencyProvider::new(dependency_provider.clone(), 50_000), name, ver);
+        }
+    }
+
+    #[test]
+    /// This test is mostly for profiling
+    fn prop_passes_int(
+        (dependency_provider, cases) in registry_strategy(0u16..665, 666)
+    )  {
+        for (name, ver) in cases {
+            let _ = resolve(&TimeoutDependencyProvider::new(dependency_provider.clone(), 50_000), name, ver);
+        }
+    }
+
+    #[test]
+    /// This tests wheter the allgorithm is still deterministic
+    fn prop_same_on_repeated_runs(
+        (dependency_provider, cases) in registry_strategy(0u16..665, 666)
+    )  {
+        for (name, ver) in cases {
+            let one = resolve(&TimeoutDependencyProvider::new(dependency_provider.clone(), 50_000), name, ver);
+            for _ in 0..3 {
+                match (&one, &resolve(&TimeoutDependencyProvider::new(dependency_provider.clone(), 50_000), name, ver)) {
```

Comment:
Should we check that the difference is not due to the timeout?

Reply:
Maybe, In Cargo we do not have fine grained control over the timeout. So I have not yet thought about how to take advantage of the better controls here.